api/internalutils/stream/stream.go

```diff
@@ -372,3 +372,97 @@ func RateLimit[T any](stream Stream[T], wait func() error) Stream[T] {
+// The streams can be of different types, however, the resulting merged stream must be of one type.
+type compareStreams[T, U, V any] struct {
+	streamA Stream[T]
+	streamB Stream[U]
+	itemA   T
+	itemB   U
+	// hasItemA keeps track of whether we have an item available from streamA.
+	// We use this flag instead of just checking whether itemA is nil in order to ensure that this adapter also works with non-nullable types.
+	hasItemA bool
+	// hasItemB keeps track of whether we have an item available from streamB.
+	hasItemB bool
+	compare  func(a T, b U) bool
+	// convert is the function that is used to convert an item from the streamB type into the V type which is the type of the merged stream.
+	// If the streams are of the same type, this can simply be a function that returns the item as-is.
+	// convertA converts an item from streamA into the V type.
+	convertA func(item T) V
+	// convertB converts an item from streamB into the V type.
+	convertB func(item U) V
+}
+
+// Next attempts to advance each stream to the next item. If false is returned, then no more items are available.
+func (cs *compareStreams[T, U, V]) Next() bool {
+	// Attempt to advance to the next item in streamA, if we don't already have an item from it.
+	if !cs.hasItemA && cs.streamA.Next() {
+		cs.itemA = cs.streamA.Item()
+		cs.hasItemA = true
+	}
+	// Attempt to advance to the next item in streamB, if we don't already  have an item from it.
+	if !cs.hasItemB && cs.streamB.Next() {
+		cs.itemB = cs.streamB.Item()
+		cs.hasItemB = true
+	}
+
+	// Return true if either stream has an item available.
+	return cs.hasItemA || cs.hasItemB
+}
+
+// Item yields the current item.
+// If only an item from one stream is available, then it will yield that item.
+// If the items from both streams are available, then the compare function will be used to decide which item to yield.
+// After yielding an item, it will also reset the availability flag of the yielded item's stream so that subsequent calls know to fetch a new item.
+func (cs *compareStreams[T, U, V]) Item() V {
+	// If both streams have items available, use the compare function to determine which to yield.
+	if cs.hasItemA && cs.hasItemB {
+		if cs.compare(cs.itemA, cs.itemB) {
+			// Reset the hasItemA flag since it has been consumed.
+			cs.hasItemA = false
+			return cs.convertA(cs.itemA)
+		} else {
+			// Reset the hasItemB flag since it has been consumed.
+			cs.hasItemB = false
+			return cs.convertB(cs.itemB)
+		}
+	}
+
+	// If only streamA has an item available, yield it.
+	if cs.hasItemA {
+		cs.hasItemA = false
+		return cs.convertA(cs.itemA)
+	}
+
+	// If only streamB has an item available, yield it.
+	if cs.hasItemB {
+		cs.hasItemB = false
+		return cs.convertB(cs.itemB)
+	}
+
+	// If neither stream has an available item, then this function should not have been called at all and there is a logic error.
+	panic("Item() was called but neither stream has an item, this is a bug")
+}
+
+// Done closes both streams.
```

Comment:
Does it though? If you get an error on `streamA` you don't even attempt `streamB`.